runtime/complex.go

```diff
@@ -169,6 +178,185 @@ func complexCoerce(o *Object) (complex128, bool) {
 	return complex(floatO, 0.0), true
 }
 
+func complexNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 0 {
+		return newObject(t), nil
+	}
+	if argc > 2 {
+		return nil, f.RaiseType(TypeErrorType, "'__new__' of 'complex' requires at most 2 arguments")
+	}
+	if t != ComplexType {
+		// Allocate a plain complex then copy it's value into an object
+		// of the complex subtype.
+		x, raised := complexNew(f, ComplexType, args, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		result := toComplexUnsafe(newObject(t))
+		result.value = toComplexUnsafe(x).Value()
+		return result.ToObject(), nil
+	}
+	o := args[0]
+	if complexSlot := o.typ.slots.Complex; complexSlot != nil && argc == 1 {
+		c, raised := complexConvert(complexSlot, f, o)
+		if raised != nil {
+			return nil, raised
+		}
+		return c.ToObject(), nil
+	}
+	if o.isInstance(StrType) {
+		if argc == 2 {
+			return nil, f.RaiseType(TypeErrorType, "complex() can't take second arg if first is a string")
+		}
+		s := toStrUnsafe(o).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+	if argc == 2 && args[1].isInstance(StrType) {
+		return nil, f.RaiseType(TypeErrorType, "complex() second arg can't be a string")
+	}
+	var cr complex128
+	crIsComplex := false
+	if complexSlot := o.typ.slots.Complex; complexSlot != nil {
+		c, raised := complexConvert(complexSlot, f, o)
+		if raised != nil {
+			return nil, raised
+		}
+		cr = c.Value()
+		crIsComplex = true
+	} else if floatSlot := o.typ.slots.Float; floatSlot != nil {
+		result, raised := floatConvert(floatSlot, f, o)
+		if raised != nil {
+			return nil, raised
+		}
+		cr = complex(result.Value(), 0)
+	} else {
+		return nil, f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+	}
+	var ci complex128
+	ciIsComplex := false
+	if argc == 1 {
+		ci = complex(imag(cr), 0)
+	} else if complexSlot := args[1].typ.slots.Complex; complexSlot != nil {
+		c, raised := complexConvert(complexSlot, f, args[1])
+		if raised != nil {
+			return nil, raised
+		}
+		ci = c.Value()
+		ciIsComplex = true
+	} else if floatSlot := args[1].typ.slots.Float; floatSlot != nil {
+		result, raised := floatConvert(floatSlot, f, args[1])
+		if raised != nil {
+			return nil, raised
+		}
+		ci = complex(result.Value(), 0)
+	} else {
+		return nil, f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+	}
+	if ciIsComplex {
```

Comment:
Rather than doing all this bookkeeping, could you just do something like this:

```
cr, raised := complexConvert(args[0])
if raised != nil {
  return nil, raised
}
var ci complex128
if len(args) > 1 {
  if ci, raised = complexConvert(args[1]); raised != nil {
    return nil, raised
  }
}
return NewComplex(cr + ci * 1i).toObject(), nil
```